File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
//         int[] chars = new int[128];
        
//         int left = 0;
//         int right = 0;
        
//         int res = 0;
//         while (right < s.length()) {
//             char s = s.charAt(right);
//             chars[r]++;
            
//             while (chars[r] > 1) {
//                 char l = s.charAt(left);
//                 chars[l]--;
//                 left++;
//             }
            
//             res = Math.max(res, right - left + 1);
            
//             right++;
//         }
//         return res;
        
        int n = s.length(), ans = 0;
        Map<Character, Integer> map = new HashMap<>();
        
        for (int j = 0, i = 0; j<n; j++){
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)), i);
            }
            ans = Math.max(ans, j - i + 1);
            map.put(s.charAt(j), j + 1);
        }
        return ans;
    }
}

File: longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s : str) -> int:
        chars = [0] * 128
        left = right = 0
        
        res = 0
        while right < len(s):
            r = s[right]
            chars[ord(r)] += 1
            
            while chars[ord(r)] > 1:
                l = s[left]
                chars[ord(l)] -= 1
                left += 1
                
            res = max(res, right -left + 1)
            
            right += 1
            
        return res



# class Solution:
#     def lengthOfLongestSubstring(self, s: str) -> int:
#         def check(start, end):
#             chars = [0] * 128
#             for i in range(start, end + 1):
#                 c = s[i]
#                 chars[ord(c)] += 1
#                 if chars[ord(c)] > 1:
#                     return False
#             return True
            
#         n = len(s)
        
#         res = 0
#         for i in range(n):
#             for j in range(i ,n):
#                 if check(i, j):
#                     res = max(res, j - i + 1)
#         return res